Add function to remap categorical labelings

bdpedigo opened the pull request:
#### Reference Issues/PRs
Fixes #526 

#### What does this implement/fix? Explain your changes.
Adds a function to try to remap labelings from one categorical labeling vector based on another. E.g. if one is clustering and you want the labels of predicted clusters to match some known truth. Has come up several times for clustering tutorials (in progress) and I also use this a decent amount in my own work so I think we should have it.

Here's a notebook with more explanation and a demo https://docs.neurodata.io/notebooks/pedigo/graspologic/clustering/2020/10/12/remapping-labels.html

#### Any other comments?
Would make more sense to put in sklearn someday, if they want it.

bdpedigo commented:
@j1c or @asaadeldin11 could you have a look for JHU side?
@Nyecarr or @dwaynepryce or @bryantower could one of you have a look for MSR side?

build-documentation breaking looks to be something in dev since I've seen in several open PRs, looks like something with intersphinx breaking. microsoft checks also stalled on this PR.

dwaynepryce commented on graspologic/utils/utils.py:
```diff
@@ -701,3 +705,51 @@ def plug_in_variance_estimator(x):
+
+def remap_labels(y_true, y_pred, return_map=False):
+    """
+    Attempts to remap a categorical labeling (such as one predicted by a clustering
```
I would reword this - attempts is too nebulous and can imply a variety of things you may not intend.  If you do intend attempts, I'd describe the optimal circumstance as well as an example or examples of the negative cases too where we didn't succeed in automatically relabeling.

bdpedigo replied:
Addressed. I expanded a little bit at the top and added a `Notes` section. If you want more than what is in the notes currently, I think it'd be easier at that point to just make an actual tutorial. I thought that was overkill for this function but I do already have a notebook linked at the top.

dwaynepryce commented on graspologic/utils/utils.py:
```diff
@@ -8,7 +8,11 @@
 import numpy as np
+from scipy.optimize import linear_sum_assignment
+from sklearn.metrics import confusion_matrix
+from sklearn.metrics._classification import _check_targets
```
Since this is a private function in sklearn, we should copy it out into our own code base or rewrite it entirely (depending on licensing details).  We can't guarantee they'll maintain this and relying/using it in our own code will make us more fragile.

bdpedigo replied:
fair. their license is BSD-3 ([link](https://github.com/scikit-learn/scikit-learn/blob/master/COPYING)). Do you know the answer for that license? If not, I am not sure how to find out with legal.

bdpedigo replied:
@dwaynepryce or @carolyncb can you advise on this point?

bdpedigo replied:
update: just need to include in license file -@carolyncb

dwaynepryce replied:
@carolyncb - still think we need to run this by CELA, since we're inlining their code.  pretty sure we just have to maintain a header in the file, but what does this do to the overarching LICENSE.txt for the project? we should just get guidance.

j1c commented on graspologic/utils/utils.py:
```diff
@@ -701,3 +705,51 @@ def plug_in_variance_estimator(x):
     return plug_in_variance_estimator
+
+
+def remap_labels(y_true, y_pred, return_map=False):
```
Add type hints. This should be pretty simple for this function.

See [this](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html) for a quick guide.

bdpedigo replied:
done. though I need to check mypy output. do you know how to do typing for variable output possibilities? I looked around and couldn't find anything. not sure if it's important, I think the main use case is just getting the array back in this PR. 

also, this code should technically work on anything sklearn considers "array-like" which I assume is a massive list of stuff. For now I just put list, array, series since again, I think this is 99.9% of use cases

dwaynepryce replied:
Union[type1, type2, type3]

dwaynepryce replied:
https://docs.python.org/3/library/typing.html#typing.Union

bdpedigo replied:
@dwaynepryce i was just talking about what stuff goes in the Union[]

bdpedigo replied:
since "array-like" is vague and I'm sure a lot of stuff would actually work

bdpedigo commented:
@dwaynepryce @j1c any further comments?

bdpedigo commented:
@asaadeldin11 @j1c @dwaynepryce would one of you have time to review this, I think it is all done as far as I know. I've addressed:
- typing
- removed dependency on sklearn private code

would like to get this in as #491 #516 #539 all need it, I think.

bdpedigo commented on graspologic/utils/utils.py:
```diff
@@ -701,3 +706,90 @@ def plug_in_variance_estimator(x):
+    Remaps a categorical labeling (such as one predicted by a clustering algorithm) to
+    match the labels used by another similar labeling.
+
+    Given two $n$-length vectors describing a categorical labeling of $n$ samples, this
```
I will fix the $$ and make math

asaadeldin11 commented:
i can review!

bdpedigo commented:
@asaadeldin11 thanks, should be good to go whenever you are

asaadeldin11 commented on graspologic/utils/utils.py:
```diff
@@ -701,3 +706,89 @@ def plug_in_variance_estimator(x):
+
+    Given two :math:`n`-length vectors describing a categorical labeling of :math:`n`
+    samples, this method reorders the labels of the second vector (`y_pred`) so that as
+    many samplesas possible from the two label vectors are in the same category.
```
```suggestion
    many samples as possible from the two label vectors are in the same category.
```

asaadeldin11 commented on graspologic/utils/utils.py:
```diff
@@ -701,3 +706,89 @@ def plug_in_variance_estimator(x):
+
+    valid_target_types = {"binary", "multiclass"}
+    if (true_type not in valid_target_types) or (pred_type not in valid_target_types):
+        msg = "Elements of `y_true` and `y_pred` must represent a valid binary or "
+        msg += "multiclass labeling, see "
```
```suggestion
        msg = ("Elements of `y_true` and `y_pred` must represent a valid binary or "
              "multiclass labeling, see ")
```
Believe the multiline string can also be done like this. Both achieve the same thing, so whether or not you want to change is up to you.

asaadeldin11 reviewed (approved):
Few minor comments, overall looks good.
